core/src/main/java/hudson/model/Slave.java

```diff
@@ -341,14 +345,23 @@ private void _setLabelString(String labelString) {
         this.labelAtomSet = Collections.unmodifiableSet(Label.parse(label));
     }
 
-    @NonNull
+    @CheckForNull // should be @NonNull, but we've seen plugins overriding readResolve() without calling super.
     private transient Set<LabelAtom> labelAtomSet;
 
     @Override
     protected Set<LabelAtom> getLabelAtomSet() {
+        if (labelAtomSet == null) {
+            warnPlugin();
+            this.labelAtomSet = Collections.unmodifiableSet(Label.parse(label));
+        }
         return labelAtomSet;
     }
 
+    private void warnPlugin() {
+        LOGGER.log(Level.WARNING, () -> getClass().getName() + " or one of its superclass overrides readResolve() without calling super implementation." +
+                "Please file an issue against the plugin implementing it: " + Jenkins.get().getPluginManager().whichPlugin(getClass()));
```

Comment:
This could also happen if an agent config hasn't been saved since the plugin has been updated, right? If so, should it be mentioned ("after applying an update, make sure to save the agent configuration" or similar)? Or do we not know enough about how these implementations work to give actionable advice?

Reply:
Node gets persisted as soon as it is added to Jenkins via `Jenkins#addNode`, so I don't think an explicit save is required.
The problem here was prominent for `labelAtomSet` because this is a transient field, and it is supposed to be computed when creating a `Slave` or after deserializing it (via `readResolve()`).

Reply:
Thanks, I was unaware they're saved immediately after being loaded.